### PR TITLE
TG-838 Documentation of the string solver

### PR DESCRIPTION
**Pull request by romainbrenguier**
This is based on https://github.com/diffblue/cbmc/pull/1420 and documents the new version of the solver.
This adds a README.md file in `solvers/refinement` which produce a page `String solver` in `Modules` of the Doxygen manual of cprover.

### CODE REVIEW

File: src/solvers/refinement/README.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  * `i = cprover_primitive(args)` where `i` is of signed bit vector type.
+    String primitives are listed in the next section.
+
+\note In the implementation, equations that are not of these forms are passed 
+  to an embedded `bv_refinementt` solver.
+
+\subsection string-representation String representation in the solver
+
+String primitives take pointers to characters for arguments representing 
+strings. To each of these pointers the string solver associate a char array
```
Comment (allredj) [inline src/solvers/refinement/README.md]: First sentence is not very clear. Maybe split into two.

File: src/solvers/refinement/string_constraint_generator_constants.cpp [hunk 2]
```diff
@@ -15,9 +15,10 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
 #include <util/prefix.h>
 #include <util/unicode.h>
 
-/// add axioms saying the returned string expression should be equal to the
+/// Add axioms saying the returned string expression should be equal to the
```
Comment (LAJW) [inline src/solvers/refinement/string_constraint_generator_constants.cpp]: I would rephrase this as: "Add axioms saying that provided string expression and constant are equal and return the expression".

File: src/solvers/refinement/string_constraint_generator_comparison.cpp [hunk 3]
```diff
@@ -13,11 +13,16 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
 
 #include <solvers/refinement/string_constraint_generator.h>
 
-/// add axioms stating that the result is true exactly when the strings
-/// represented by the arguments are equal. the variable ending in
-/// `witness_unequal` is -1 if the length differs or an index at which the
-/// strings are different
-/// \par parameters: function application with two string arguments
+/// Equality of the content of two strings
+///
```
Comment (LAJW) [inline src/solvers/refinement/string_constraint_generator_comparison.cpp]: I feel like those "code section descriptions" shouldn't be a part of the function comment.

It would make sense to denote them with double slash and/or suggest that this code section should go into a separate file. But that's just a suggestion.

File: src/solvers/refinement/string_constraint_generator_indexof.cpp [hunk 4]
```diff
@@ -316,11 +319,14 @@ exprt string_constraint_generatort::add_axioms_for_last_index_of(
...
-/// add axioms corresponding to the String.lastIndexOf:(C),
+/// Index of the last occurence of a target inside the string (slow for target
+/// longer than one character)
+///
+/// Add axioms corresponding to the String.lastIndexOf:(C),
 /// String.lastIndexOf:(CI), String.lastIndexOf:(String), and
 /// String.lastIndexOf:(String,I) java functions
-/// \par parameters: function application with 2 or 3 arguments
-/// \return a integer expression
+/// \param f: function application with 2 or 3 arguments
```
Comment (LAJW) [inline src/solvers/refinement/string_constraint_generator_indexof.cpp]: This could be changed to "Add axioms for an index of the last occurrence..." and "Supports following overloads:"

**Review by LAJW — APPROVED** [review]
I'd say that this documentation looks good enough.

**Comment by allredj** [comment]
@romainbrenguier Could you squash the documentation commits please so that they are easier to review?

**Comment by romainbrenguier** [comment]
@allredj 
> @romainbrenguier Could you squash the documentation commits please so that they are easier to review?

Done, the commits specific to this PR are the two last ones.

**Review by thk123 — APPROVED** [review]
I have reviewed the changes in src/pointer-analysis *only* but they look fine.

File: src/solvers/refinement/README.md [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+which represents the content of the string.
+If the pointer is the address of an actual array in the program they should be 
+linked by using the primitive `cprover_string_associate_array_to_pointer`.
+The length of the array can also be linked to a variable of the program using
+ `cprover_string_associate_length_to_array`.
+
+\warning The solver assumes the memory pointed by the arguments is immutable
+which is not something that is true in general for C pointers for instance.
+Therefore for each transformation on a string, it is assumed the program 
+allocates new string before calling a primitive.
```
Comment (allredj) [inline src/solvers/refinement/README.md]: `new string` -> `a new string`

**Review by allredj — APPROVED** [review]
I didn't read everything in detail, but it looks fine.

File: src/solvers/refinement/string_constraint_generator_indexof.cpp [hunk 6]
```diff
@@ -225,10 +241,23 @@ exprt string_constraint_generatort::add_axioms_for_last_index_of_string(
...
 }
 
-/// add axioms corresponding to the String.indexOf:(C), String.indexOf:(CI),
-/// String.indexOf:(String), and String.indexOf:(String,I) java functions
-/// \par parameters: function application with 2 or 3 arguments
-/// \return a integer expression
+/// Index of the first occurence of a target inside the string
+///
+/// If the target is a character:
+/// \copybrief add_axioms_for_index_of(const array_string_exprt&,const exprt&,const exprt&)
```
Comment (peterschrammel) [inline src/solvers/refinement/string_constraint_generator_indexof.cpp]: Line too long. Please fix http://cbmc-linter.diffblue.com.s3-website-us-east-1.amazonaws.com/linter-b07fcddda78d14adffe806abd4891189902d7eb0.txt

**Review by peterschrammel — CHANGES_REQUESTED** [review]

File: src/solvers/refinement/string_constraint_generator_indexof.cpp (hunk 6 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_constraint_generator_indexof.cpp]: @peterschrammel I'm not sure how to fix this, breaking the line breaks the doxygen link, and adding a `// NOLINT` comment will make `NOLINT` appear in the doxygen documentation.

Comment (thk123) [reply src/solvers/refinement/string_constraint_generator_indexof.cpp]: Does the `NOLINTNEXTLINE` show up in the doxygen? I'd strongly prefer a error in the linter to confusion in the documentation...

Comment (romainbrenguier) [reply src/solvers/refinement/string_constraint_generator_indexof.cpp]: No because there are only two `/`s (not three) at the beginning of the line. From what I've seen, the output is fine.

**Review by peterschrammel — APPROVED** [review]

**Comment by peterschrammel** [comment]
As a doc-only change this looks good to me.
